--- a/kernel/bpf/verifier.c
+++ b/kernel/bpf/verifier.c
@@ -13491,7 +13491,7 @@ static struct bpf_prog *inline_bpf_loop(
 	/* Note: insn_buf[12] is an offset of BPF_CALL_REL instruction */
 	call_insn_offset = position + 12;
 	callback_offset = callback_start - call_insn_offset - 1;
-	env->prog->insnsi[call_insn_offset].imm = callback_offset;
+	new_prog->insnsi[call_insn_offset].imm = callback_offset;
 
 	return new_prog;
 }
