From 8aa4eb8092e67fa7c3d764b199b8a1a833b40d89 Mon Sep 17 00:00:00 2001
From: Serge Belyshev <belyshev@depni.sinp.msu.ru>
Date: Sat, 26 Oct 2019 17:08:56 +0300
Subject: [PATCH] Work around x86-only llc stuff

---
 kernel/sched/MuQSS.c | 6 ++++++
 1 file changed, 6 insertions(+)

--- a/kernel/sched/MuQSS.c
+++ b/kernel/sched/MuQSS.c
@@ -6755,7 +6755,11 @@ static bool cache_cpu_idle(struct rq *rq
 /* MC siblings CPU mask which share the same LLC */
 static const cpumask_t *llc_core_cpumask(int cpu)
 {
+#ifdef CONFIG_X86
 	return per_cpu(cpu_llc_shared_map, cpu);
+#else
+	return topology_core_cpumask(cpu);
+#endif
 }
 #endif
 
@@ -7065,6 +7069,7 @@ void __init sched_init_smp(void)
 		}
 	}
 
+#ifdef CONFIG_X86
 	for_each_online_cpu(cpu) {
 		rq = cpu_rq(cpu);
 		for (i = 0; i < total_runqueues; i++) {
@@ -7080,6 +7085,7 @@ void __init sched_init_smp(void)
 			       rq->cpu_order[i]->cpu, per_cpu(cpu_llc_id, rq->cpu_order[i]->cpu));
 		}
 	}
+#endif
 
 	switch (rqshare) {
 		case RQSHARE_ALL:
