diff --git a/applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/zones.js b/applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/zones.js
index f07df06..6e2e0f8 100644
--- a/applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/zones.js
+++ b/applications/luci-app-firewall/htdocs/luci-static/resources/view/firewall/zones.js
@@ -57,6 +57,8 @@ return view.extend({
 		};
 
 		o = s.option(form.Flag, 'drop_invalid', _('Drop invalid packets'));
+ 		o = s.option(form.Flag, 'fullcone', _('Enable FullCone NAT'));
+ 		o = s.option(form.Flag, 'fullcone6', _('Enable FullCone NAT6'));
 
 		var p = [
 			s.option(form.ListValue, 'input', _('Input')),
