diff --git a/cgroupfs-mount b/cgroupfs-mount
index 40810ba..f4c9f45 100755
--- a/cgroupfs-mount
+++ b/cgroupfs-mount
@@ -26,7 +26,11 @@ fi
 
 # mount /sys/fs/cgroup if not already done
 if ! mountpoint -q /sys/fs/cgroup; then
-	mount -t tmpfs -o uid=0,gid=0,mode=0755 cgroup /sys/fs/cgroup
+	if [ "$1" = "v2" ]; then
+		mount -t cgroup2 cgroup2 /sys/fs/cgroup
+	else
+		mount -t tmpfs -o uid=0,gid=0,mode=0755 cgroup /sys/fs/cgroup
+	fi
 fi
 
 cd /sys/fs/cgroup
@@ -34,9 +38,11 @@ cd /sys/fs/cgroup
 # get/mount list of enabled cgroup controllers
 for sys in $(awk '!/^#/ { if ($4 == 1) print $1 }' /proc/cgroups); do
 	mkdir -p $sys
-	if ! mountpoint -q $sys; then
-		if ! mount -n -t cgroup -o $sys cgroup $sys; then
-			rmdir $sys || true
+	if [ "$1" != "v2" ]; then
+		if ! mountpoint -q $sys; then
+			if ! mount -n -t cgroup -o $sys cgroup $sys; then
+				rmdir $sys || true
+			fi
 		fi
 	fi
 done
