diff --git a/package/new/r8168/src/r8168_n.c b/package/new/r8168/src/r8168_n.c
index 0cf2053a9f..ef904dc374 100644
--- a/package/new/r8168/src/r8168_n.c
+++ b/package/new/r8168/src/r8168_n.c
@@ -24441,6 +24441,7 @@ rtl8168_init_software_variable(struct net_device *dev)
                 tp->NotWrMcuPatchCode = TRUE;
         }
 
+        RTL_W16(tp, CustomLED, 0x870);
         tp->NicCustLedValue = RTL_R16(tp, CustomLED);
 
         rtl8168_get_hw_wol(dev);
