--- a/drivers/char/lrng/lrng_sysctl.c
+++ b/drivers/char/lrng/lrng_sysctl.c
@@ -86,7 +86,7 @@ void lrng_sysctl_update_max_write_thresh
 	mb();
 }
 
-static struct ctl_table random_table[] = {
+struct ctl_table random_table[] = {
 	{
 		.procname	= "poolsize",
 		.maxlen		= sizeof(int),
@@ -131,9 +131,3 @@ static struct ctl_table random_table[] =
 	},
 };
 
-static int __init random_sysctls_init(void)
-{
-	register_sysctl_init("kernel/random", random_table);
-	return 0;
-}
-device_initcall(random_sysctls_init);
