--- a/crypto/jitterentropy.c
+++ b/crypto/jitterentropy.c
@@ -655,12 +655,6 @@ struct rand_data *jent_lrng_entropy_coll
 		.memblocks	= JENT_MEMORY_BLOCKSIZE,
 		.memblocksize	= JENT_MEMORY_BLOCKS,
 		.memaccessloops	= JENT_MEMORY_ACCESSLOOPS,
-		.rct_count	= 0,
-		.apt_observations = 0,
-		.apt_count	= 0,
-		.apt_base	= 0,
-		.apt_base_set	= 0,
-		.health_failure = 0
 	};
 
 	if (jent_entropy_init())
