--- linux-5.15/drivers/char/lrng/lrng_sysctl.c.orig	2022-05-17 08:35:09.958660325 +0200
+++ linux-5.15/drivers/char/lrng/lrng_sysctl.c	2022-05-17 08:36:17.829818950 +0200
@@ -91,7 +91,7 @@ void lrng_sysctl_update_max_write_thresh
 	mb();
 }
 
-static struct ctl_table random_table[] = {
+struct ctl_table random_table[] = {
 	{
 		.procname	= "poolsize",
 		.maxlen		= sizeof(int),
@@ -137,9 +137,3 @@ static struct ctl_table random_table[] =
 	{ }
 };
 
-static int __init random_sysctls_init(void)
-{
-	register_sysctl_init("kernel/random", random_table);
-	return 0;
-}
-device_initcall(random_sysctls_init);
