--- linux-5.4/include/uapi/linux/random.h.lrng	2020-08-13 20:42:05.900054630 -0600
+++ linux-5.4/include/uapi/linux/random.h	2020-08-13 20:43:53.375216798 -0600
@@ -49,8 +49,10 @@
  *
  * GRND_NONBLOCK	Don't block and return EAGAIN instead
  * GRND_RANDOM		Use the /dev/random pool instead of /dev/urandom
+ * GRND_INSECURE	Return non-cryptographic random bytes
  */
 #define GRND_NONBLOCK	0x0001
 #define GRND_RANDOM	0x0002
+#define GRND_INSECURE	0x0004
 
 #endif /* _UAPI_LINUX_RANDOM_H */
