--- a/drivers/char/lrng/lrng_drng_chacha20.h
+++ b/drivers/char/lrng/lrng_drng_chacha20.h
@@ -28,7 +28,11 @@ struct chacha20_state {
 
 static inline void lrng_cc20_init_rfc7539(struct chacha20_block *chacha20)
 {
-	chacha_init_consts(chacha20->constants);
+	/* String "expand 32-byte k" */
+	chacha20->constants[0] = 0x61707865;
+	chacha20->constants[1] = 0x3320646e;
+	chacha20->constants[2] = 0x79622d32;
+	chacha20->constants[3] = 0x6b206574;
 }
 
 #define LRNG_CC20_INIT_RFC7539(x) \
