--- linux-5.4-2.orig/drivers/char/lrng/lrng_interfaces.c	2020-08-20 12:14:25.593091918 +0200
+++ linux-5.4-2/drivers/char/lrng/lrng_interfaces.c	2020-08-20 12:14:55.877453575 +0200
@@ -609,7 +609,6 @@ const struct file_operations random_fops
 	.write = lrng_drng_write,
 	.poll  = lrng_random_poll,
 	.unlocked_ioctl = lrng_ioctl,
-	.compat_ioctl = compat_ptr_ioctl,
 	.fasync = lrng_fasync,
 	.llseek = noop_llseek,
 };
@@ -618,7 +617,6 @@ const struct file_operations urandom_fop
 	.read  = lrng_drng_read,
 	.write = lrng_drng_write,
 	.unlocked_ioctl = lrng_ioctl,
-	.compat_ioctl = compat_ptr_ioctl,
 	.fasync = lrng_fasync,
 	.llseek = noop_llseek,
 };
