--- a/drivers/char/lrng/lrng_es_irq.c
+++ b/drivers/char/lrng/lrng_es_irq.c
@@ -657,7 +657,7 @@ static void lrng_time_process(void)
 }
 
 /* Hot code path - Callback for interrupt handler */
-void add_interrupt_randomness(int irq)
+void add_interrupt_randomness(int irq, int irq_flg)
 {
 	if (lrng_highres_timer()) {
 		lrng_time_process();
@@ -691,6 +691,7 @@ void add_interrupt_randomness(int irq)
 		tmp = lrng_raw_jiffies_entropy_store(jiffies) ? 0 : jiffies;
 		tmp ^= lrng_raw_irq_entropy_store(irq) ? 0 : irq;
 		tmp ^= lrng_raw_retip_entropy_store(ip) ? 0 : ip;
+		tmp ^= irq_flg;
 		tmp ^= ip >> 32;
 		_lrng_irq_array_add_u32(tmp);
 	}
