--- linux-6.1.14/drivers/char/lrng/lrng_interface_random_kernel.c.orig	2023-02-26 16:21:03.212109504 +0100
+++ linux-6.1.14/drivers/char/lrng/lrng_interface_random_kernel.c	2023-02-26 16:21:22.284469755 +0100
@@ -85,7 +85,7 @@ void lrng_kick_random_ready(void)
  * @entropy_bits: amount of entropy in buffer (value is in bits)
  */
 void add_hwgenerator_randomness(const void *buffer, size_t count,
-				size_t entropy_bits, bool sleep_after)
+				size_t entropy_bits)
 {
 	/*
 	 * Suspend writing if we are fully loaded with entropy or if caller
@@ -96,7 +96,6 @@ void add_hwgenerator_randomness(const vo
 	wait_event_interruptible(lrng_write_wait,
 				(lrng_need_entropy() && entropy_bits) ||
 				lrng_state_exseed_allow(lrng_noise_source_hw) ||
-				!sleep_after ||
 				kthread_should_stop());
 	lrng_state_exseed_set(lrng_noise_source_hw, false);
 	lrng_pool_insert_aux(buffer, count, entropy_bits);
