--- linux-5.4/drivers/char/lrng/lrng_es_mgr.c.orig	2020-10-10 20:56:39.331954692 +0200
+++ linux-5.4/drivers/char/lrng/lrng_es_mgr.c	2020-10-10 20:56:50.031999524 +0200
@@ -318,8 +318,8 @@ int __init rand_initialize(void)
 	seed.time = ktime_get_real();
 
 	for (i = 0; i < ARRAY_SIZE(seed.data); i++) {
-		if (!arch_get_random_seed_long_early(&(seed.data[i])) &&
-		    !arch_get_random_long_early(&seed.data[i]))
+		if (!arch_get_random_seed_long(&(seed.data[i])) &&
+		    !arch_get_random_long(&seed.data[i]))
 			seed.data[i] = random_get_entropy();
 	}
 	memcpy(&seed.utsname, utsname(), sizeof(*(utsname())));
