diff -rNEZbwBdu3 a/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts
--- a/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts	2020-09-04 00:42:48.504000000 +0800
+++ b/arch/arm64/boot/dts/rockchip/rk3328-nanopi-r2s.dts	2020-09-04 00:44:34.096000000 +0800
@@ -142,6 +142,7 @@
 	snps,reset-delays-us = <0 10000 50000>;
 	tx_delay = <0x24>;
 	rx_delay = <0x18>;
+	handle_cpu_id = <1>;
 	status = "okay";
 
 	mdio {
@@ -358,4 +359,5 @@
 &usbdrd_dwc3 {
 	dr_mode = "host";
 	status = "okay";
+	handle_cpu_id = <2>;
 };

